### PR TITLE
Add ScreenShot Preview Block

### PR DESCRIPTION
**Pull request by StevenDufresne**
This block was created to add pattern previews to the theme directory in [this Pull Request](https://github.com/WordPress/wordpress.org/pull/88). 

## Block Summary:
- Loads a visual _card_ container
- Requests thumbnail from [mShots](https://github.com/Automattic/mShots) service
- If the thumbnail is cached in the service, convert to base64, show image.
- If not, wait, re-request until the thumbnail is ready from `mShots` service.

### CODE REVIEW

**Comment by ryelle** [comment]
Rather than create an entire new project (config files, package.json, etc), could you try out the build process & template I proposed in https://github.com/WordPress/wporg-mu-plugins/pull/211 & https://github.com/WordPress/wporg-repo-tools/pull/21?

**Comment by StevenDufresne** [comment]
Yep makes sense. This block was created in a galaxy far far away...I didn't know it would end up here :)

**Comment by StevenDufresne** [comment]
I would like to get this out as a beta on .org, so I'll merge this now and follow up later. If there are any issues, I'll follow them up asap.

File: mu-plugins/blocks/screenshot-preview/block.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Plugin name: Gutenberg: Pattern Previewer
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/block.php:3]: There's a bit of inconsistency between the naming and description here and in `block.json`. This file is quite particular about it being for previewing patterns, whereas `block.json` seems more generic. Is that intentional?

File: mu-plugins/blocks/screenshot-preview/block.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * Description: A block that displays a pattern.
+ * Version:     2.0
+ * Author:      WordPress.org
+ * Author URI:  http://wordpress.org/
+ * License:     GPLv2 or later
+ */
+
+namespace WordPressdotorg\Gutenberg\ScreenshotPreview;
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/block.php:11]: `Gutenberg` -> `MU_Plugins` ?

File: mu-plugins/blocks/screenshot-preview/js/block.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+				setFrameHeight( getCardFrameHeight( wrapperRef.current.clientWidth ) );
+			} catch ( err ) {}
+		};
+
+		handleOnResize();
+
+		window.addEventListener( 'resize', handleOnResize );
+
+		return () => {
+			window.addEventListener( 'resize', handleOnResize );
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/js/block.js:31]: If this is intended to clean up should it be using `removeEventListener`?;

File: mu-plugins/blocks/screenshot-preview/js/block.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	}, [ isVisible ] );
+
+	return (
+		<a
+			className="wporg-screenshot-card"
+			ref={ wrapperRef }
+			style={ {
+				height: frameHeight,
+			} }
+			tabIndex="-1"
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/js/block.js:48]: Is this to stop keyboard navigation? If so why?

File: mu-plugins/blocks/screenshot-preview/package.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    "format": "wp-scripts format",
+    "lint:css": "wp-scripts lint-style",
+    "lint:js": "wp-scripts lint-js",
+    "start": "wp-scripts start js/index.js"
+  },
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "@wordpress/scripts": "^23.4.0",
+    "lodash": "^4.17.21"
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/package.json:17]: Looks like this should be in `dependencies`

File: mu-plugins/blocks/screenshot-preview/js/in-view.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/js/in-view.js:9]: I don't see this package in package.json `dependencies`

File: mu-plugins/blocks/screenshot-preview/js/index.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import Block from './block.js';
+
+const init = () => {
+	const blockElements = document.getElementsByClassName( 'wporg-screenshot-preview-js' );
+
+	if ( ! blockElements ) {
+		return;
+	}
+
+	for ( let i = 0; i < blockElements.length; i++ ) {
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/js/index.js:18]: Any reason not to use `forEach`?

File: mu-plugins/blocks/screenshot-preview/js/screenshot.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useEffect, useRef, useState } from '@wordpress/element';
+import { Spinner } from '@wordpress/components';
```
Comment (adamwoodnz) [inline mu-plugins/blocks/screenshot-preview/js/screenshot.js:6]: Missing from package.json `dependencies`?

**Review by adamwoodnz — COMMENTED** [review]
Nice work, this is a cool block!

A few minor comments/questions inline, I may be missing some context/know-how though, if so sorry!

File: mu-plugins/blocks/screenshot-preview/js/block.js (hunk 3 above)
Comment (StevenDufresne) [reply mu-plugins/blocks/screenshot-preview/js/block.js:31]: Great catch!

File: mu-plugins/blocks/screenshot-preview/js/index.js (hunk 7 above)
Comment (StevenDufresne) [reply mu-plugins/blocks/screenshot-preview/js/index.js:18]: What are the arguments for using `forEach`?

File: mu-plugins/blocks/screenshot-preview/block.php (hunk 2 above)
Comment (StevenDufresne) [reply mu-plugins/blocks/screenshot-preview/block.php:11]: 👍 Yes!

File: mu-plugins/blocks/screenshot-preview/block.php (hunk 1 above)
Comment (StevenDufresne) [reply mu-plugins/blocks/screenshot-preview/block.php:3]: Nope, just missed this. Nice catch.

File: mu-plugins/blocks/screenshot-preview/js/index.js (hunk 7 above)
Comment (adamwoodnz) [reply mu-plugins/blocks/screenshot-preview/js/index.js:18]: This pretty much sums it up https://stackoverflow.com/questions/43031988/javascript-efficiency-for-vs-foreach#answer-43032526

Personally as a dev I much prefer to write and read less code. I don't think the perf benefit with a for loop outweighs this in most cases. I tend to stick to all the modern native array functions: forEach, map, some, reduce

**Pull request merged**